dotnet/src/Connectors/Connectors.MistralAI/Client/MistralClient.cs

```diff
@@ -33,52 +34,89 @@ internal sealed class MistralClient
-        var request = this.CreateChatCompletionRequest(modelId, false, chatHistory, mistralExecutionSettings);
-        var endpoint = new Uri($"{this._endpoint}{this._separator}chat/completions");
+        var request = this.CreateChatCompletionRequest(modelId, stream: false, chatHistory, mistralExecutionSettings);
+        var endpoint = this.GetEndpoint(mistralExecutionSettings, path: "chat/completions");
 
         using var httpRequestMessage = this.CreatePost(request, endpoint, this._apiKey, stream: false);
 
         var response = await this.SendRequestAsync<ChatCompletionResponse>(httpRequestMessage, cancellationToken).ConfigureAwait(false);
 
-        return response.Choices.Select(chatChoice => new ChatMessageContent(new AuthorRole(chatChoice.Message!.Role), chatChoice.Message!.Content, this._modelId, chatChoice, Encoding.UTF8, GetChatChoiceMetadata(response, chatChoice))).ToList();
+        return response.Choices.Select(chatChoice => new ChatMessageContent(new AuthorRole(chatChoice.Message!.Role!), chatChoice.Message!.Content, this._modelId, chatChoice, Encoding.UTF8, GetChatChoiceMetadata(response, chatChoice))).ToList();
     }
 
-    internal async IAsyncEnumerable<StreamingChatMessageContent> GetStreamingChatMessageContentsAsync(ChatHistory chatHistory, CancellationToken cancellationToken, PromptExecutionSettings? executionSettings = null, Kernel? kernel = null)
+    internal async IAsyncEnumerable<StreamingChatMessageContent> GetStreamingChatMessageContentsAsync(ChatHistory chatHistory, [EnumeratorCancellation] CancellationToken cancellationToken, PromptExecutionSettings? executionSettings = null, Kernel? kernel = null)
     {
         string modelId = executionSettings?.ModelId ?? this._modelId;
         var mistralExecutionSettings = MistralAIPromptExecutionSettings.FromExecutionSettings(executionSettings);
-        var request = this.CreateChatCompletionRequest(modelId, false, chatHistory, mistralExecutionSettings);
-        var endpoint = new Uri($"{this._endpoint}{this._separator}chat/completions");
+        var request = this.CreateChatCompletionRequest(modelId, stream: true, chatHistory, mistralExecutionSettings);
+        var endpoint = this.GetEndpoint(mistralExecutionSettings, path: "chat/completions");
 
         using var httpRequestMessage = this.CreatePost(request, endpoint, this._apiKey, stream: true);
 
         using var response = await this.SendStreamingRequestAsync(httpRequestMessage, cancellationToken).ConfigureAwait(false);
 
         using var responseStream = await response.Content.ReadAsStreamAndTranslateExceptionAsync().ConfigureAwait(false);
-        foreach (var streamingTextContent in this.ProcessResponseStream(responseStream, modelId))
+        using var reader = new StreamReader(responseStream);
+        string line;
+        string? rawChunk = null;
+        AuthorRole? currentRole = null;
+        while ((line = await reader.ReadLineAsync().ConfigureAwait(false)) != null)
         {
```

Comment:
If you want, you can reuse my updated version of StreamJsonParser from gemini.
After update it is fully async and still fast.
https://github.com/microsoft/semantic-kernel/blob/feature-connectors-gemini/dotnet/src/Connectors/Connectors.GoogleVertexAI/Core/StreamJsonParser.cs

Reply:
The way MistralAI outputs the chunks seems to be easier to Deserialize using a simpler approach for skipping the `line breaks`, @markwallace-microsoft may worth considering if the `IStreamJsonParser` interface can be applied to your streaming logic, this then can be another Parser utility we can use across different connectors.

Reply:
Is it just using server-sent events?

Reply:
@stephentoub  Yes, here's the extract from the documentation

_Whether to stream back partial progress. If set, tokens will be sent as data-only server-sent events as they become available, with the stream terminated by a data: [DONE] message. Otherwise, the server will hold the request open until the timeout or until completion, with the response containing the full result as JSON._

Reply:
@Krzysztof318 @RogerBarreto My understanding of the stream parser is it's used to identify each chunk my checking for completed JSON. In the Mistral case I think it's simpler because each chunk is delimited with an empty line.

Reply:
@Krzysztof318 , might worth having a specialized JsonParser for SSE, this [DONE] is something similar to what OpenAI do with theirs... BTW, SSE is a pattern. 😊 (SSEStreamParser)

Reply:
Created this issue to track: https://github.com/microsoft/semantic-kernel/issues/5610